File: problems/guess-who/solution/GuessWho.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class GuessWho {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Input parsing
        int testCases = Integer.parseInt(scanner.nextLine().trim());
        List<String> results = new ArrayList<>();

        for (int t = 0; t < testCases; t++) {
            int numberOfGuests = Integer.parseInt(scanner.nextLine().trim());
            List<String[]> guests = new ArrayList<>();

            for (int i = 0; i < numberOfGuests; i++) {
                guests.add(scanner.nextLine().trim().split(" "));
            }

            String[] clues = scanner.nextLine().trim().split(" ");
            List<String> matchingGuests = findMysteryGuest(guests, clues);

            if (matchingGuests.isEmpty()) {
                results.add("No match found.");
            } else {
                results.add(String.join(" ", matchingGuests));
            }
        }

        // Output results
        for (String result : results) {
            System.out.println(result);
        }

        scanner.close();
    }

    public static List<String> findMysteryGuest(List<String[]> guests, String[] clues) {
        List<String> matchingGuests = new ArrayList<>();

        for (String[] guest : guests) {
            String name = guest[0];
            List<String> traits = List.of(guest).subList(1, guest.length);

            if (traits.contains(clues[0]) && traits.contains(clues[1])) {
                matchingGuests.add(name);
            }
        }

        return matchingGuests;
    }
}

File: problems/G-guess-who/solution/guess-who.py
def find_mystery_guest(number_of_guests, guests, clues):
    matching_guests = []
    for guest in guests:
        name, traits = guest[0], set(guest[1:])
        if all(clue in traits for clue in clues):
            matching_guests.append(name)
    
    if len(matching_guests) != 0:
        return matching_guests
    else:
        return "No match found."

# Input parsing
test_cases = int(input().strip())
results = []

for _ in range(test_cases):
    number_of_guests = int(input().strip())
    guests = [input().strip().split() for _ in range(number_of_guests)]
    clues = input().strip().split()
    result = find_mystery_guest(number_of_guests, guests, clues)
    results.append(result)

# Output results
for result in results:
    if isinstance(result, list): # if theres multiple people
        print(*result)
    else:
        print(result)